### PR TITLE
Invert serializer from user default to protocol extension

### PR DESCRIPTION
**Pull request by basememara**
This was an idea just for discussion and to see any thoughts on this.. moving the logic from the `UserDefaults` extension to a `UserDefaultsSerializable` protocol extension. The idea was to let the serializer implement the divergence in handling instead putting the burden on the `UserDefaults`.

An example of this could be seen with the one-off `URL` handling where it's in the protocol override instead of in the fetch logic. The trick was to enforce a static protocol conformance on the type and let the `UserDefaults` be injected into the static function:

```swift
public protocol UserDefaultsSerializable {
    associatedtype StoredValue

    var storedValue: StoredValue { get }

    init(storedValue: StoredValue)

    static func fetch(forKey defaultName: String, from userDefaults: UserDefaults) -> Self
    static func fetchOptional(forKey defaultName: String, from userDefaults: UserDefaults) -> Self?
    static func set(_ value: Self, forKey defaultName: String, from userDefaults: UserDefaults)
}
```

Then the default conformance for all serializers:

```swift
extension UserDefaultsSerializable {
    public var storedValue: Self { self }

    public init(storedValue: Self) {
        self = storedValue
    }

    public static func fetch(forKey defaultName: String, from userDefaults: UserDefaults) -> Self {
        fetchOptional(forKey: defaultName, from: userDefaults)!
    }

    public static func fetchOptional(forKey defaultName: String, from userDefaults: UserDefaults) -> Self? {
        guard let object = userDefaults.object(forKey: defaultName) as? StoredValue else { return nil }
        return Self(storedValue: object)
    }

    public static func set(_ value: Self, forKey defaultName: String, from userDefaults: UserDefaults) {
        userDefaults.set(value.storedValue, forKey: defaultName)
    }
}
```

Then one-off's like `URL` could be done in the protocol override conformances instead of in the `UserDefaults` extension:
```swift
extension URL: UserDefaultsSerializable {
    public static func fetchOptional(forKey defaultName: String, from userDefaults: UserDefaults) -> Self? {
        userDefaults.url(forKey: defaultName)
    }

    public static func set(_ value: Self, forKey defaultName: String, from userDefaults: UserDefaults) {
        userDefaults.set(value, forKey: defaultName)
    }
}
```

This swaps the property wrapper logic from a `UserDefaults` extension to the type's protocol extension:
```swift
@propertyWrapper
public struct WrappedDefault<T: UserDefaultsSerializable> {
    private let userDefaults: UserDefaults
    public let key: String

    public var wrappedValue: T {
        get { T.fetch(forKey: key, from: userDefaults) }
        set { T.set(newValue, forKey: key, from: userDefaults) }
    }

    public init(
        forKey key: String,
        defaultValue: T,
        userDefaults: UserDefaults = .standard
    ) {
        self.key = key
        self.userDefaults = userDefaults
        self.userDefaults.register(defaults: [key: defaultValue.storedValue])
    }
}
```

The tests still pass and added more, but wondering if this is a good direction or any ideas on how to break this.

### CODE REVIEW

File: Tests/IntegrationTests.swift [hunk 1]
```diff
@@ -94,19 +111,20 @@ final class IntegrationTests: XCTestCase {
 
         let newValue = "text data".data(using: .utf8)
         settings.data = newValue
-        XCTAssertEqual(TestSettings.store.fetch("data"), newValue)
+        XCTAssertEqual(TestSettings.store.data(forKey: "data"), newValue)
 
         settings.data = nil
-        XCTAssertNil(TestSettings.store.fetchOptional("data") as URL?)
```
Comment (basememara) [inline Tests/IntegrationTests.swift:100]: If PR rejected, this typo should be fixed and casted to `Data?` not `URL` otherwise.

Comment (jessesquires) [reply Tests/IntegrationTests.swift:100]: Opps, thanks! A separate fix would be great here. 😊

**Comment by jessesquires** [comment]
Thanks for exploring this design @basememara 👍🏼 💯 

It's an interesting idea, but I find the inversion a bit awkward (because of the `static func`). It is slightly more burdensome for conformers of `UserDefaultsSerializable` too. 

I like the nicer approach for handling `URL` types, but I do not think this change is worth it. If every type required a unique implementation, then I would find this approach more appealing. But the existing one-off hack for `URL` is ok with me since it is hidden away. 😊 

If anything, I would be more inclined to make the `UserDefaults` extensions public instead of the changes here, but I don't see a strong case (or demand from users) for that right now.

**Comment by jessesquires** [comment]
Oh also, I noticed you opened this from a fork.

As a `collaborator`, you can work directly from a branch in this repo now if you like. 😊 The PR workflow is a bit nicer that way.

**Comment by basememara** [comment]
Nice thx for giving it some thought. The `static` didn't sit well with me either, especially on a protocol. 

The protocol extension may still work as a default implementation even without this change, so the conformity is optional, but still testing what the implications are of using `Self` which I believe is determined at runtime so it's ok, but also the implicit implementation maybe confusing or seem magic a bit.

Great feedback thanks. I'll open a separate PR for the typo in the test.

**Pull request closed without merging**